src/renderer/components/opds/BrowserResult.tsx

```diff
@@ -16,82 +16,98 @@ import {
-        if (prevProps.url !== this.props.url || prevProps.location.search !== this.props.location.search) {
+        if (prevProps.url !== this.props.url ||
+            prevProps.location.search !== this.props.location.search) {
             // New url to browse
             this.browseOpds();
         }
     }
 
-    public render(): React.ReactElement<{}>  {
-        const { result } = this.props;
-        let content = (<Loader/>);
-        if (result && result.isFailure) {
-            // browse error
-        } else if (result) {
-            switch (result.data.type) {
-                case OpdsResultType.NavigationFeed:
-                    content = (
-                        <EntryList entries={ result.data.navigation } />
-                    );
-                    break;
-                case OpdsResultType.PublicationFeed:
-                    content = (
-                        <EntryPublicationList publications={ result.data.publications } />
-                    );
-                    break;
-                case OpdsResultType.Empty:
-                    // TRANSLATE
-                    content = (
-                        <Empty />
-                    );
-                    break;
-                default:
-                    break;
+    public render(): React.ReactElement<{}> {
+        const { result, resultIsReject, __ } = this.props;
+        let content = (<Loader />);
+        if (typeof result === "string" && resultIsReject) {
+            content = (
+                <MessageOpdBrowserResult title={__("opds.network.reject")} message={result} />
+            );
+        } else if (typeof result === "object" && result) {
+            if (result.isSuccess) {
+                switch (result.data.type) {
+                    case OpdsResultType.NavigationFeed:
+                        content = (
+                            <EntryList entries={result.data.navigation} />
+                        );
+                        break;
+                    case OpdsResultType.PublicationFeed:
+                        content = (
+                            <EntryPublicationList publications={result.data.publications} />
+                        );
+                        break;
+                    case OpdsResultType.Empty:
+                        content = (
+                            <MessageOpdBrowserResult title={__("opds.empty")} />
+                        );
+                        break;
+                    default:
+                        break;
+                }
+            } else if (result.isTimeout) {
+                content = (
+                    <MessageOpdBrowserResult title={__("opds.network.timeout")} />
+                );
+            } else {
+                content = (
+                    <MessageOpdBrowserResult
+                        title={`${__("opds.network.error")}${result.statusCode} ${result.statusMessage}`}
+                    />
+                );
```

Comment:
`typeof result === "object" && resultIsReject` is not captured by these conditional statements. Is that intentional? (I assume that a promise rejection / unexpected error signaled with `resultIsReject` can originate from a situation where `result` is not a string ... but I must admit not being totally aware of the details under the hood)

Reply:
i handle it now,

Reply:
So just to be clear, based on the current code:

1) if `resultIsReject` is `true` (equivalent to `ApiAction.type === ActionType.Error`, see `apiReducer()` function), and:
2) `result` is non-falsy (i.e. non-null/undefined `ApiAction.payload`), and:
3) the type of `result` is `"object"` (i.e. not `string`), then:
4) `MessageOpdBrowserResult` is invoked with `result.statusCode` / `result.statusMessage` (which can be `undefined` => bug)

If I am not mistaken, the only condition that is protecting us from state (4) is the fact that we can assume that step (3) never yields true in practice, because rejections when `ActionType.Error` usually mean that `ApiAction.payload` has type `string`. But I don't like assumptions like this, especially for the sake of ` future code maintenance.

So, I would like to suggest that `if (resultIsReject)` should be moved **before** `if (typeof result === "object" && result)` to ensure the rejection is handled first, irrespective of the type of `result` (which is why we stringify it, because it could hypothetically be `any`-thing).

(please let me know if you think my concern is unfounded, maybe I am missing something)

Reply:
When resultIsReject is true the result props must be a string.

I applied your suggest https://github.com/readium/readium-desktop/pull/542/commits/2b7e93baa8484816e1d8cd024a6b5fdb215b8d3f